---
title: "Plot metagenomics data - Auckland Run2 - 16S"
author: "Pierre-Yves Dupont"
date: "July 27, 2018"
output:
  html_document:
    css: my_css.css
    highlight: default
    number_sections: no
    theme: cerulean
    toc: yes
    toc_depth: 2
    toc_float:
      collapsed: no
      smooth_scroll: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Libraries

```{r libs, warning=FALSE, message=F}
library(phyloseq); packageVersion("phyloseq")
library(tidyverse)
library(caching); packageVersion("caching")
library(cowplot)
library(assertthat)
library(ggplot2)
library(RColorBrewer)
library(colorRamps)
library(ggrepel)
require(car)
library(ggpubr)
```

## Loading data

The objects computed using DADA2 have been saved using the `caching` library and can be loaded using the same library. Here we can load the `phyloseq` object `ps` and the `data.frame` extracted from it `taxa.dt`.

```{r load, warning=F, message=F}
ps <- load.object("psc")$orig
if(object.cached("taxa.dt")){
  taxa.dt <- load.object("taxa.dt")
} else{
  taxa.dt <- psmelt(ps)
  save.object(taxa.dt)
}
```
## Loading Metadata

```{r, message=F, warning=F}
metadata <- read_tsv("samples-metadata.tsv")
metadata.df <- as.data.frame(metadata)
rownames(metadata.df) <- metadata$SampleID
sample_data(ps) <- metadata.df
if(!exists("taxa.dt")){
  taxa.dt <- psmelt(ps)
  save.object(taxa.dt)
}
```


## Separate UC samples

```{r}
uc_samples <- c("GW180247","GW180419","GW180420","GW180421","GW180422","GW180423","GW180424","GW180425","GW180426","GW180427","GW180428", "GW1801011A", "GW180248")
ps_uc <- prune_samples(sample_data(ps)$SampleID %in% uc_samples, ps)
```

## Diversity

### Filter low abundance taxa.

```{r}
aggregation.level <- 'Genus'

dt <- taxa.dt %>% filter(Blank == "No", SampleID %in% uc_samples)
top <- dt %>% 
  group_by(UQ(rlang::sym(aggregation.level))) %>% 
  summarise(Abundance=sum(Abundance)) %>% 
  mutate(AbundanceRel = Abundance/sum(Abundance)) %>%
  filter(AbundanceRel >= 0.01)
```

#### Compute relative abundance 

```{r}
dt.grouped <- dt %>%
  filter(UQ(rlang::sym(aggregation.level)) %in% top$UQ(rlang::sym(aggregation.level)))  %>%
  mutate(UQ(rlang::sym(aggregation.level)) := factor(UQ(rlang::sym(aggregation.level)), levels=unique(UQ(rlang::sym(aggregation.level))))) %>%
  filter(!is.na(UQ(rlang::sym(aggregation.level)))) %>% # Remove NA
  group_by(SampleID, UQ(rlang::sym(aggregation.level))) %>%
  summarize(Abundance = sum(Abundance)) %>%
  mutate(Abundance = Abundance/sum(Abundance) * 100) %>%
  filter(!is.na(UQ(rlang::sym(aggregation.level))))
```

#### Set the palette

```{r}
colourCount <- length(unique(as.character(dt.grouped[[aggregation.level]])))
palette <- colorRampPalette(brewer.pal(10,"Paired"))(colourCount)
```

##### Plot

###### All samples

```{r, fig.width=8, fig.height=8}
ggplot(dt.grouped, aes(SampleID, Abundance, fill = UQ(rlang::sym(aggregation.level)))) +
  geom_bar(stat = "identity", position = "stack") +
  # facet_wrap(~SampleTypeAnalysed, ncol = 1, scales = "free_x") +
  theme_minimal() +
  scale_fill_manual(values = palette) +
  xlab(NULL) +
  ylab("Relative abundance") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank()) +
  theme(legend.position="bottom")
```

###### Only 3 samples

The following code builds a plot using the same taxa set as for the complete plot by subsetting the `dt.grouped` data.frame. This doesn't change the distribution of the taxa.

```{r, fig.width=8, fig.height=8}
dt.grouped <- dt.grouped %>% filter(SampleID %in% c("GW180247", "GW180419", "GW180421"))
ggplot(dt.grouped, aes(SampleID, Abundance, fill = UQ(rlang::sym(aggregation.level)))) +
  geom_bar(stat = "identity", position = "stack") +
  # facet_wrap(~SampleTypeAnalysed, ncol = 1, scales = "free_x") +
  theme_minimal() +
  scale_fill_manual(values = palette) +
  xlab(NULL) +
  ylab("Relative abundance") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank()) +
  theme(legend.position="bottom")
```

The following code first subset the data then recomputes the taxa and build the abundance plot.

```{r}

dt <- taxa.dt %>%
  filter(Blank == "No", SampleID %in% uc_samples) %>% 
  filter(SampleID %in% c("GW180247", "GW180419", "GW180421"))

top <- dt %>% 
  group_by(UQ(rlang::sym(aggregation.level))) %>% 
  summarise(Abundance=sum(Abundance)) %>% 
  mutate(AbundanceRel = Abundance/sum(Abundance)) %>%
  filter(AbundanceRel >= 0.01)

dt.grouped <- dt %>%
  filter(UQ(rlang::sym(aggregation.level)) %in% top$UQ(rlang::sym(aggregation.level)))  %>%
  mutate(UQ(rlang::sym(aggregation.level)) := factor(UQ(rlang::sym(aggregation.level)), levels=unique(UQ(rlang::sym(aggregation.level))))) %>%
  filter(!is.na(UQ(rlang::sym(aggregation.level)))) %>% # Remove NA
  group_by(SampleID, UQ(rlang::sym(aggregation.level))) %>%
  summarize(Abundance = sum(Abundance)) %>%
  mutate(Abundance = Abundance/sum(Abundance) * 100) %>%
  filter(!is.na(UQ(rlang::sym(aggregation.level))))

colourCount <- length(unique(as.character(dt.grouped[[aggregation.level]])))
palette <- colorRampPalette(brewer.pal(10,"Paired"))(colourCount)
```

```{r, fig.width=8, fig.height=8}
ggplot(dt.grouped, aes(SampleID, Abundance, fill = UQ(rlang::sym(aggregation.level)))) +
  geom_bar(stat = "identity", position = "stack") +
  # facet_wrap(~SampleTypeAnalysed, ncol = 1, scales = "free_x") +
  theme_minimal() +
  scale_fill_manual(values = palette) +
  xlab(NULL) +
  ylab("Relative abundance") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.background = element_blank()) +
  theme(legend.position="bottom")
```

Both plots are correct. The first one corresponds to the distribution of the taxa identified as the most abundant in all the samples while the second corresponds to the distribution of the the most abundant taxa in these three samples. It is not surpising to see that you have such a large difference between the plots because your conditions (and likely your communities) are very different. 